cou/steps/hypervisor.py

```diff
@@ -160,6 +160,31 @@ def get_azs(self, target: OpenStackRelease) -> AZs:
 
         return azs
 
+    def _upgrade_plan_sanity_checks(
+        self, target: OpenStackRelease, group: HypervisorGroup
+    ) -> None:
+        """Run sanity checks before generating upgrade plan for hypervisor group.
+
+        :param target: OpenStack codename to upgrade.
+        :type target: OpenStackRelease
+        :param group: HypervisorGroup object
+        :type group: HypervisorGroup
+        """
+        for app in self.apps:
+            if app.name not in group.app_units:
+                logger.debug(
+                    "skipping application %s because it is not part of group %s",
+                    app.name,
+                    group.name,
+                )
+                continue
+
+            units = group.app_units[app.name]
+            logger.info("running sanoty checks for %s units of %s app", app.name, units)
+            # Note(rgildein): We don't catch the error here because we shouldn't generate any
+            #                 update plan if sanity checks for any application in the group fails.
+            app.upgrade_plan_sanity_checks(target, units)
```

Comment:
With the current implementation, hypervisor planning will stop on the first application which fails its sanity check. While this is ok, we miss the opportunity to provide information to user if multiple applications are having issues (e.g. both `nova-compute` and `cinder` have problems but the user is only made aware of one). I think it would be helpful to finish planning for all applications, at least within the same AZ, so the operator can be better informed.

Reply:
Maybe, but as we agreed it's not important now, and it's safer to simply quit. Because if Cinder have issue, you should not update nova-compute and same for az, if one az failed you should not update next one.

Reply:
As discussed during the sync, the current implementation makes sense if we have unit-level sanity checks and we want to continue generating plan for other AZs even if one fails. Since the conclusion is that we will not do either of thses, I think it would be better do sanity checks at once at the higher level for each application with their available-to-upgrade units (across all AZs) considered by HypervisorPlanner. I see three benefits of doing so:
1. visually clear that we are doing sanity checks on application level
2. avoid calling the same sanity checks multiple times (once per AZ in the current implementation)
3. with one less for-loop (right now we have two, one for AZ and one for applications), it would be easier to catch exceptions for multiple applications and print them all (though we agreed that this is not a requirement).

This being stated, I wouldn't block this PR just because of this if you feel strongly about keeping the current approach.

Reply:
Posting a note after our re-discussion.

> sanity checks are designed to be run by unit level

Not really with what we have right now. We are accepting units as arguments in sanity checks but we don't have any checks yet for units themselves. But things will change after #365 being addressed. So I'm +1 on reverting back to doing sanity checks per AZ. I appreciate all these discussions :slightly_smiling_face: